### DISCUSSION TITLE
Music Playback Roadmap

### DISCUSSION THREAD

**Discussion by 1hitsong** [jellyfin-archive/jellyfin-roku-legacy#608] [General]
Here's a general direction I'm planning to take with music library playback. They are roughly in order (top down) that I'll work them.

Items covered by MVP PR
- [x] MVP - Basic functionality to get single song and album playback working. 
- [x] Improve NowPlaying layout and design - Now that things are working, let's make it look better. This includes using standard playback icons instead of buttons
- [x] Add loop playback mode
- [x] Add shuffle playback mode
- [x] Add current spot to NowPlaying
- [x] Instant Mix support - https://github.com/jellyfin/jellyfin-roku/pull/642
- [x] Add Album View - https://github.com/jellyfin/jellyfin-roku/pull/640
- [x] When music is playing, change screensaver to display audio information (Cover art, artist/album/song info)
- [x] Add Artist View - https://github.com/jellyfin/jellyfin-roku/pull/640
- [x] Add Genre View - https://github.com/jellyfin/jellyfin-roku/pull/881
- [x] Add total song length stamp to NowPlaying
- [x] Add music to search results (Artists, Albums, Songs) - https://github.com/jellyfin/jellyfin-roku/pull/593

Next Steps
- [ ] When clicking on a song, load album as playlist and start playback at selected song's index
- [ ] Add Song View
- [ ] Write Tests for MVP functionality - If test suite is prepared, otherwise push back
- [ ] Improve NowPlaying data fallbacks - Try every avenue to get data for screen
- [ ] See if possible to add song progress bar & allow user to move through song using it

**Comment by cewert** [comment]
> Write Tests for MVP functionality - If test suite is prepared, otherwise push back

I don't think any of the testing stuff is working. At least it wasn't last i tried #35. It's a great idea but you might want to make that it's own project haha

**Reply by 1hitsong** [reply]
Cool. I'll poke around a bit and see if it's usable or not. If it's not, I'll at least see if it's a quick fix to get it up and running. If it's a big lift, I agree, let's make it its own project/task.

**Comment by cewert** [comment]
> - [x]  Add current spot and total song length stamp to NowPlaying

I don't see any information about song length or current position in my current testing. When was this implemented? I searched but I couldn't find a PR
![PXL_20230319_151145691 MP](https://user-images.githubusercontent.com/16514652/226439062-862370ce-26a3-4891-ba1e-ee50044c1854.jpg)

**Reply by 1hitsong** [reply]
I believe current spot is referring to where you are in the current playback queue. In your screenshot, it would be Track 1/1.

Song length stamp isn't implemented.

**Reply by 1hitsong** [reply]
I may have been referring to the playback bar too. Originally, there was no bar.

**Reply by cewert** [reply]
Gotcha. I was reading that as the total song time and current time in song.

Can we add current time in song to your list somewhere? I don't know if I'm using the right words to describe it so here's a picture from spotify to show what I mean. I want to add both numbers shown
![spotify-player-ui](https://user-images.githubusercontent.com/16514652/226479550-234cf16e-22e8-4e87-a449-04ad3b736403.PNG)

**Reply by 1hitsong** [reply]
Yeah man. We can certainly do that.

**Comment by vollink** [comment]
Hi there.  Dumb question, maybe, but I see playlist referred to as something to do with an Album, but I don't see Playlist as a stand-alone thing I can select and play.  I don't see the ability to use playlists in the existing Roku client, and I don't see it on the roadmap.  Am I missing something, or is this a feature nobody is actually asking for?

**Reply by cewert** [reply]
@vollink Playlist support was added to the roku client in #986 and first appeared in `v.1.6.4`.

Playlists are currently user specific - only the user who created the playlist will be able to view the playlist on the roku. This will change in a future server release to allow users to share access to other users. For more info read through this issue - #1318.

If that doesn't solve your problem, please create a new issue so we can troubleshoot and figure out what's going on.
